Assignment 06 - SENTIMENT ANALYSIS

Natural language processing has extensively applied in sentiment analysis as
a significant challenge. In this scenario, the objective is to determine if the
tweets shared by customers regarding technology companies that produce
and sell mobiles, computers, laptops, and similar products express positive
sentiment or negative sentiment. The goal will be to build a system that can
accurately classify the new tweets sentiments. You can divide the data into
train and test. The Evaluation metric you should use is the accuracy.


In [36]:
# load dataset
import pandas as pd
import io
df = pd.read_csv("/content/tweets.csv")

In [37]:
# display data
df

,id,label,tweet
0,1,0,#fingerprint #Pregnancy Test https://goo.gl/h1...
1,2,0,Finally a transparant silicon case ^^ Thanks t...
2,3,0,We love this! Would you go? #talk #makememorie...
3,4,0,I'm wired I know I'm George I was made that wa...
4,5,1,What amazing service! Apple won't even talk to...
...,...,...,...
7915,7916,0,Live out loud #lol #liveoutloud #selfie #smile...
7916,7917,0,We would like to wish you an amazing day! Make...
7917,7918,0,Helping my lovely 90 year old neighbor with he...
7918,7919,0,Finally got my #smart #pocket #wifi stay conne...


In [38]:
# Import libraries
import pandas as pd
import numpy as np
import re
import nltk

from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [39]:
nltk.download('wordnet')
nltk.download('stopwords')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

Clean Text

In [40]:
# clean tweets
stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

def preprocess(text):
    text = text.lower()
    text = re.sub(r"http\S+|@\w+|#\w+", "", text)   # remove URLs, mentions
    text = re.sub(r"[^a-z\s]", "", text)            # remove punctuation

    words = text.split()
    words = [lemmatizer.lemmatize(w) for w in words if w not in stop_words]

    return " ".join(words)

df['tweet'] = df['tweet'].apply(preprocess)

In [41]:
df

,id,label,tweet
0,1,0,test
1,2,0,finally transparant silicon case thanks uncle
2,3,0,love would go
3,4,0,im wired know im george made way
4,5,1,amazing service apple wont even talk question ...
...,...,...,...
7915,7916,0,live loud
7916,7917,0,would like wish amazing day make every minute ...
7917,7918,0,helping lovely year old neighbor ipad morning ...
7918,7919,0,finally got stay connected anytimeanywhere


TF-IDF

In [42]:
vectorizer = TfidfVectorizer(max_features=5000)

X = vectorizer.fit_transform(df['tweet'])
y = df['label']

Split data into training and testing

In [43]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [44]:
X_train.shape

(5544, 5000)

In [45]:
X_test.shape

(2376, 5000)

In [46]:
y_train.shape

(5544,)

In [47]:
y_test.shape

(2376,)

Build Model

In [48]:
# Logistic Regression model
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

In [49]:
#Prediction
y_pred = model.predict(X_test)

Evaluation

In [50]:
# Accuracy
print("Accuracy:", accuracy_score(y_test, y_pred))

Accuracy: 0.8333333333333334


In [51]:
#Classification Report
print("\nClassification Report:\n", classification_report(y_test, y_pred))


Classification Report:
               precision    recall  f1-score   support

           0       0.83      0.97      0.89      1735
           1       0.84      0.48      0.61       641

    accuracy                           0.83      2376
   macro avg       0.83      0.72      0.75      2376
weighted avg       0.83      0.83      0.82      2376



In [52]:
#Confusion Matrix
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))


Confusion Matrix:
 [[1675   60]
 [ 336  305]]


The Logistic Regression model achieved an accuracy of 83.33%, indicating that it can correctly classify most tweets as positive or negative.